MH-13265, Automatic caption using Google speech to text api

tyfranck opened the pull request:
Implementation of automatic captions/transcriptions using Google speech to text. 
Please review :-)

tyfranck committed:
MH-13265 Automatic caption using Google speech to text api - implementation

lkiesow commented:
Quick first glance:

- test failures
- database migration script is missing

lkiesow commented on modules/transcription-service-google-speech-impl/src/main/resources/OSGI-INF/google-speech-transcription-rest.xml:
```diff
@@ -0,0 +1,22 @@
+
+	<service>
+		<provide
+			interface="org.opencastproject.transcription.googlespeech.endpoint.GoogleSpeechTranscriptionRestService" />
```
indentation/tabs

lkiesow commented on modules/transcription-service-google-speech-impl/src/main/resources/OSGI-INF/google-speech-transcription-service.xml:
```diff
@@ -0,0 +1,39 @@
+	name="org.opencastproject.transcription.googlespeech.GoogleSpeechTranscriptionService"
+	activate="activate" immediate="true">
+	<implementation
+		class="org.opencastproject.transcription.googlespeech.GoogleSpeechTranscriptionService" />
```
tabs

tyfranck commented:
@lkiesow what do you mean by 'database migration script is missing'. 
I have added my table scripts to 'mysql5.sql', is that not enough?

tyfranck committed:
MH-13265 fixed tabs/indentations following comments

tyfranck committed:
MH-13265 removed empty spaces to fix test failures

lkiesow commented:
That's for new installations.
There are also the migration scripts: `docs/upgrade/6_to_7/mysql5.sqldocs/upgrade/6_to_7/mysql5.sql`

tyfranck committed:
MH-13265 added missing database migration scripts following comments

smarquard commented:
Just noting here that the persistent service here is suitable for use by multiple services. We have an implementation of a transcription service for Nibity (https://opencast.jira.com/browse/MH-13308) that we will submit in due course, and it will use these generalized persistence modules (so we don't have to have new tables for every new transcription or captioning service).

gregorydlogan commented on docs/guides/admin/docs/modules/googlespeechtranscripts.md:
```diff
@@ -0,0 +1,251 @@
+Overview
+--------
+
+The GoogleSpeechTranscriptionService invokes the Google Speech-to-Text service via REST API to translate audio to text.
```
`translate` should perhaps be `transcribe`?

gregorydlogan commented on docs/guides/admin/docs/modules/googlespeechtranscripts.md:
```diff
@@ -0,0 +1,251 @@
+During the execution of an Opencast workflow, an audio file is extracted from one of the presenter videos and sent 
+to the Google Speech-to-Text service. When the results are received, they are converted to the desired caption format 
+and attached to the media package.
+
```
To me it was unclear that this is an async operation - at least from a naive user's point of view.  This could be clarified with something like:

`Because Google's Speech-to-Text service can take a while to process a recording, we do not wait for it to finish before proceeding with the rest of Opencast's normal processing.  Logically, this looks something like this:`

tyfranck replied:
fixed

gregorydlogan commented on docs/guides/admin/docs/modules/googlespeechtranscripts.md:
```diff
@@ -0,0 +1,251 @@
+Configuration
+-------------
+
+### Step 1: Activate Google Speech and Google Cloud Storage APIs
```
So, my experience writing the docs for the AWS integration tells me that by the time someone's actually reading these docs the UI will have changed *drastically*, or been removed completely.  A warning here saying that this is how it's currently working and to consult Google's documentation for $keywords is probably warranted.

tyfranck replied:
I've added a note

gregorydlogan commented on docs/guides/admin/docs/modules/googlespeechtranscripts.md:
````diff
@@ -0,0 +1,251 @@
+email notifications will be disabled. 
+
+```
+# Change enabled to true to enable this service. 
````
This example of the file should either match the instructions above, or be removed IMO.

tyfranck replied:
fixed

gregorydlogan commented on docs/guides/admin/docs/modules/googlespeechtranscripts.md:
```diff
@@ -0,0 +1,251 @@
+### Step 4: Add encoding profile for extracting audio
+
+The Google Speech-to-Text service has limitations on audio types. [Supported audio type are here](https://cloud.google.com/speech-to-text/docs/reference/rest/v1p1beta1/RecognitionConfig#AudioEncoding).
+Try using the encoding profile suggested in etc/encoding/googlespeech-audio.properties.
```
Doesn't this PR include those settings?  If they're in the encodings directory there's no action required.

`Try using...` -> `By default Opencast will use the encoding settings in $file`

tyfranck replied:
fixed

gregorydlogan commented on docs/guides/admin/docs/modules/googlespeechtranscripts.md:
```diff
@@ -0,0 +1,251 @@
+    </operation>
+
+    <!-- Merge caption/transcript to existing publication and republish -->
+
```
Is there an operation missing here?  If not, this comment should be removed.

gregorydlogan commented on docs/guides/admin/docs/workflowoperationhandlers/google-speech-attach-transcription-woh.md:
```diff
@@ -0,0 +1,63 @@
+      description="Attach captions/transcription">
+      <configurations>
+        <!-- This is filled out by the transcription service when starting this workflow -->
+        <configuration key="transcription-job-id">${transcriptionJobId}</configuration>
```
If this parameter's value can't change without breaking things, why are we exposing it to the user?  Is there a usecase where changing this value is useful?

tyfranck replied:
This workflow is not accessible by user via admin. Only admin can access it via workflow directory.  And if they know what they're doing they can set it to attach a given job id.

gregorydlogan replied:
Ok, fair enough.

gregorydlogan commented on docs/guides/admin/docs/workflowoperationhandlers/google-speech-start-transcription-woh.md:
```diff
@@ -0,0 +1,61 @@
+|source-flavor|The flavor of the audio file to be sent for translation.|EMPTY|presenter/delivery|
+|source-tag|The flavor of the audio file to be sent for translation.|EMPTY|transcript|
+|skip-if-flavor-exists|If this flavor already exists in the media package, skip this operation.<br/>To be used when the media package already has a transcript file. Optional|false|captions/timedtext|
+|language-code|The language code to use for the transcription. Optional. If set, it will override the configuration language code|EMPTY|en-US|
```
Does Google have a link for valid language codes?

tyfranck replied:
link added

gregorydlogan commented on etc/org.opencastproject.transcription.googlespeech.GoogleSpeechTranscriptionService.cfg:
```diff
@@ -0,0 +1,48 @@
+enabled=false
+
+# Google Cloud Service details
+google.cloud.client.id=
```
Config keys with default values set should be commented out.  The exception is for things like `enabled` above, which can remain uncommented.

tyfranck replied:
fixed

gregorydlogan commented on etc/workflows/google-speech-attach-transcripts.xml:
```diff
@@ -0,0 +1,79 @@
+      </configurations>
+    </operation>
+
+    <!-- Merge caption/transcript to existing publication and republish -->
```
Same as the docs above, is there a missing step here?

gregorydlogan commented on modules/caption-impl/src/main/java/org/opencastproject/caption/converters/GoogleSpeechCaptionConverter.java:
```diff
@@ -0,0 +1,188 @@
+    int transcriptionLineSize = 0;
+    try {
+      // No language to specify so define size of a transcripts line
+      transcriptionLineSize = Integer.parseInt(languageLineSize.trim());
```
Probably makes sense to log the size selected here as well.

gregorydlogan commented on modules/caption-impl/src/main/java/org/opencastproject/caption/converters/GoogleSpeechCaptionConverter.java:
```diff
@@ -0,0 +1,188 @@
+  }
+
+  private String removeCharacter(String str, char car) {
+    if (str != null && str.length() > 0 && str.charAt(str.length() - 1) == car) {
```
`if (str != null && str.endsWith(car)))` ?

tyfranck replied:
"car" for character :-) , this code is to remove last character from the provided google timestamp which has the format: 100s.  See line 117 on how it's used: (removeCharacter((wordTSList.get("endTime").toString()), 's'))
@gregorydlogan  do you want me to write it differently?

gregorydlogan replied:
@tyfranck My initial comment was my suggestion :)

You're checking if it's null, that its length is greater than zero, and then if the last character matches.  I'm checking if it's null, then if it ends with a given character.  I would even simplify my initial comment with

`if (str.endsWith(car))`, since I'm 99% sure `.endsWith` checks for nulls.

tyfranck replied:
I get it now :-) . Completely misread it.

tyfranck replied:
fixed

gregorydlogan commented on modules/transcription-service-google-speech-impl/src/main/java/org/opencastproject/transcription/googlespeech/GoogleSpeechTranscriptionService.java:
```diff
@@ -0,0 +1,1064 @@
+  private static final String PROVIDER = "Google Speech";
+
+  // Global configuration (custom.properties)
+  public static final String ADMIN_URL_PROPERTY = "org.opencastproject.admin.ui.url";
```
This key is in `OpencastConstants.java`, the rest (IMO) should be but that's kinda out of scope for this PR :)

tyfranck replied:
fixed

gregorydlogan commented on modules/transcription-service-google-speech-impl/src/main/java/org/opencastproject/transcription/googlespeech/GoogleSpeechTranscriptionService.java:
```diff
@@ -0,0 +1,1064 @@
+  private static final String DIGEST_USER_PROPERTY = "org.opencastproject.security.digest.user";
+
+  // Cluster name
+  private static final String CLUSTER_NAME_PROPERTY = "org.opencastproject.environment.name";
```
This is already in the Watson version, and should be moved up either into `OpencastConstants.java` or into a translation service constants file.

tyfranck replied:
fixed

gregorydlogan commented on modules/transcription-service-google-speech-impl/src/main/java/org/opencastproject/transcription/googlespeech/GoogleSpeechTranscriptionService.java:
```diff
@@ -0,0 +1,1064 @@
+   * Service configuration options
+   */
+  public static final String ENABLED_CONFIG = "enabled";
+  public static final String GOOGLE_SPEECH_LANGUAGE = "google.speech.language";
```
I feel like the keys with prefixed `google.speech` *really* ought to live in a Google provided jar pulled out of Maven somewhere, but I can't find them in their docs.  Something to look out for with revision 2 of this code I guess.

tyfranck replied:
yes, something to keep an eye on but doesn't currently exist

gregorydlogan commented on modules/transcription-service-google-speech-impl/src/main/java/org/opencastproject/transcription/googlespeech/GoogleSpeechTranscriptionService.java:
```diff
@@ -0,0 +1,1064 @@
+          } catch (NumberFormatException e) {
+            // Use default
+            logger.warn("Invalid configuration for {} : {}. Default used instead: {}",
+                    new Object[]{COMPLETION_CHECK_BUFFER_CONFIG, bufferOpt.get(), completionCheckBuffer});
```
I seem to remember @lkiesow went through and changed a ton of logging lines either to this style, or away from it.  Check the rest of the codebase and make sure it matches.

tyfranck replied:
@lkiesow can you please advice on the logging style @gregorydlogan is mentioning in his comment above.

tyfranck replied:
Not sure what to change here

gregorydlogan commented on modules/transcription-service-google-speech-impl/src/main/java/org/opencastproject/transcription/googlespeech/GoogleSpeechTranscriptionService.java:
```diff
@@ -0,0 +1,1064 @@
+          try {
+            maxProcessingSeconds = Long.parseLong(maxProcessingOpt.get());
+          } catch (NumberFormatException e) {
+            // Use default
```
This should be logged, since it's a configuration error (either an uncommented blank, or not numeric).

tyfranck replied:
fixed

gregorydlogan commented on modules/transcription-service-google-speech-impl/src/main/java/org/opencastproject/transcription/googlespeech/GoogleSpeechTranscriptionService.java:
```diff
@@ -0,0 +1,1064 @@
+          try {
+            cleanupResultDays = Integer.parseInt(cleaupOpt.get());
+          } catch (NumberFormatException e) {
+            // Use default
```
Logging here.

tyfranck replied:
fixed

gregorydlogan commented on modules/transcription-service-google-speech-impl/src/main/java/org/opencastproject/transcription/googlespeech/GoogleSpeechTranscriptionService.java:
```diff
@@ -0,0 +1,1064 @@
+
+        logger.info("Activated!");
+      } else {
+        logger.info("Service disabled. If you want to enable it, please update the service configuration.");
```
In general I would reverse the order of this if-else clause.

```
if (!enabled) {
  logger.info("Service disabled...
  return;
}
<the rest of the method>
```

This avoids indenting the whole thing, and clarifies the logic.

tyfranck replied:
fixed

gregorydlogan commented on modules/transcription-service-google-speech-impl/src/main/java/org/opencastproject/transcription/googlespeech/GoogleSpeechTranscriptionService.java:
```diff
@@ -0,0 +1,1064 @@
+        logger.info("Service disabled. If you want to enable it, please update the service configuration.");
+      }
+    } else {
+      throw new IllegalArgumentException("Missing component context");
```
I'm not *opposed* to throwing this here since (IMO) it is an error, but does this ever actually happen aside from testing?  IIRC, we don't check for this anywhere else, and if the CC is null something is *very* wrong anyway :)

tyfranck replied:
fixed

gregorydlogan commented on modules/transcription-service-google-speech-impl/src/main/java/org/opencastproject/transcription/googlespeech/GoogleSpeechTranscriptionService.java:
```diff
@@ -0,0 +1,1064 @@
+            logger.info("Google Cloud Service access token created");
+            return accessToken;
+          }
+          return INVALID_TOKEN;
```
I would be tempted to log an error, throw and exception here specifically saying that the 200 happened but you received bad data from Google.  This is a different case than any non-200 result, which is a (hopefully) temporary failure which should be retried.

If you're getting the invalid token back on an HTTP 200 then something is _very_ wrong.

This also has the side effect of not needing to check whether it's the invalid token everywhere you request it.  Either the token is cached (and good), or you request a new one and it's either good, or Google is broken and you throw an exception.  There's no case where you need it and it *hasn't* be checked immediately before.

tyfranck replied:
fixed

gregorydlogan commented on modules/transcription-service-google-speech-impl/src/main/java/org/opencastproject/transcription/googlespeech/GoogleSpeechTranscriptionService.java:
```diff
@@ -0,0 +1,1064 @@
+      }
+    } catch (IOException e) {
+      logger.warn("Could not save transcription results file for mpId {}, jobId {}: {}",
+              new Object[]{mpId, jobId, jsonObj == null ? "null" : jsonObj.toJSONString()});
```
Logging type again.

gregorydlogan replied:
Is there a reason this wasn't changed if you addressed the other ones?

gregorydlogan commented on modules/transcription-service-google-speech-impl/src/main/java/org/opencastproject/transcription/googlespeech/GoogleSpeechTranscriptionService.java:
```diff
@@ -0,0 +1,1064 @@
+              new Object[]{mpId, jobId, jsonObj == null ? "null" : jsonObj.toJSONString()});
+      throw new TranscriptionServiceException("Could not save transcription results file", e);
+    } catch (TranscriptionDatabaseException e) {
+      logger.warn("Transcription results file were saved but state in db not updated for mpId {}, jobId {}", mpId,
```
I'm assuming this exception is triggered by line 378 (which IMO sounds like a race condition).  If so, this sounds like something that should be ignored (or better yet, doesn't happen at all), except we throw an exception right after this.  What other cases could trigger this exception?

tyfranck replied:
yes line 378, it could be a race condition but main motivation is for example if Google storage services or its API's are inaccessible at the time of deletion, job should not be mark as completed to avoid orphan file on google storage.

gregorydlogan replied:
Ah, good point.  Ok, let's leave this then.

gregorydlogan commented on modules/transcription-service-google-speech-impl/src/main/java/org/opencastproject/transcription/googlespeech/GoogleSpeechTranscriptionService.java:
```diff
@@ -0,0 +1,1064 @@
+              String.format("There was a transcription error for for media package %s, job id %s.",
+                      jobControl.getMediaPackageId(), jobId));
+    } catch (TranscriptionDatabaseException e) {
+      logger.warn("Transcription error. State in db could not be updated to error for mpId {}, jobId {}", mpId, jobId);
```
Same comment as above, won't continue banging on this.

gregorydlogan commented on modules/transcription-service-google-speech-impl/src/main/java/org/opencastproject/transcription/googlespeech/GoogleSpeechTranscriptionService.java:
```diff
@@ -0,0 +1,1064 @@
+      return false;
+    }
+    try {
+      HttpGet httpGet = new HttpGet(GOOGLE_SPEECH_URL + RESULT_PATH + jobId);
```
`UrlSupport.concat` has the benefit of handling the `/` characters for you, thus removing them from your constants.

tyfranck replied:
fixed

gregorydlogan commented on modules/transcription-service-google-speech-impl/src/main/java/org/opencastproject/transcription/googlespeech/GoogleSpeechTranscriptionService.java:
```diff
@@ -0,0 +1,1064 @@
+    } catch (Exception e) {
+      String msg = String.format("Exception when calling the recognitions endpoint for media package %s, job id %s",
+              mpId, jobId);
+      logger.warn(String.format(msg, mpId, jobId), e);
```
`msg` is being formatted twice

tyfranck replied:
fixed

gregorydlogan commented on modules/transcription-service-google-speech-impl/src/main/java/org/opencastproject/transcription/googlespeech/GoogleSpeechTranscriptionService.java:
```diff
@@ -0,0 +1,1064 @@
+      return "No results found";
+    }
+    try {
+      HttpGet httpGet = new HttpGet(GOOGLE_SPEECH_URL + RESULT_PATH + jobId);
```
`UrlSupport` again, not going to keep banging on this so check the rest of this PR.

tyfranck replied:
fixed

gregorydlogan commented on modules/transcription-service-google-speech-impl/src/main/java/org/opencastproject/transcription/googlespeech/GoogleSpeechTranscriptionService.java:
```diff
@@ -0,0 +1,1064 @@
+  void createRecognitionsJob(String mpId, Track track, String languageCode) throws TranscriptionServiceException, IOException {
+    // Use default language if not set by workflow
+    if (languageCode == null || languageCode.isEmpty()) {
+      languageCode = language;
```
Just realized that `language` here is a service-wide default, and thus should be renamed to `defaultLangauge` or something.

tyfranck replied:
fixed

gregorydlogan commented on modules/transcription-service-google-speech-impl/src/main/java/org/opencastproject/transcription/googlespeech/GoogleSpeechTranscriptionService.java:
```diff
@@ -0,0 +1,1064 @@
+   */
+  public String getTranscriptionStatus(String mpId) throws TranscriptionServiceException {
+    try {
+      for (TranscriptionJobControl jc : database.findByMediaPackage(mpId)) {
```
So, we're returning one result from this method, but the database can contain multiple job controls.  I'm assuming we support re-transcribing the same mediapackage (say, a retry with a new/better speech model).  But don't we want to get the *latest* job control?  The underlying query here would return them in whatever the default order is for the table, which I'm guessing is actually the opposite of what you're looking for with this method.

tyfranck replied:
The underlying query order the results by date created. This is specify in  'TranscriptionJobControlDto.java' in the database module. The latest result should be the first one.

gregorydlogan commented on modules/transcription-service-google-speech-impl/src/main/java/org/opencastproject/transcription/googlespeech/GoogleSpeechTranscriptionService.java:
```diff
@@ -0,0 +1,1064 @@
+    return "Unknown";
+  }
+
+  protected CloseableHttpClient makeHttpClient() throws IOException {
```
The wrapping in this method is wonky.  I would do something like
```
RequestConfig reqConfig = RequestConfig.custom()
        .setConnectTimeout(CONNECTION_TIMEOUT)
        .setSocketTimeout(SOCKET_TIMEOUT)
        .setConnectionRequestTimeout(CONNECTION_TIMEOUT)
        .build();
return HttpClients.custom().setDefaultRequestConfig(reqConfig).build();
```

tyfranck replied:
fixed

gregorydlogan commented on modules/transcription-service-google-speech-impl/src/main/java/org/opencastproject/transcription/googlespeech/GoogleSpeechTranscriptionService.java:
```diff
@@ -0,0 +1,1064 @@
+
+    try {
+      HttpPost httpPost = new HttpPost(tokenEndpoint + String.format(
+              "?client_id=%s&client_secret=%s&refresh_token=%s&grant_type=refresh_token",
```
UrlSupport probably has bits for query params.

tyfranck replied:
Nope, UrlSupport doesn't seem to support query params

gregorydlogan commented on modules/transcription-service-google-speech-impl/src/main/java/org/opencastproject/transcription/googlespeech/GoogleSpeechTranscriptionService.java:
```diff
@@ -0,0 +1,1064 @@
+      JSONObject jsonObject = (JSONObject) jsonParser.parse(jsonString);
+      switch (code) {
+        case HttpStatus.SC_OK: // 200
+          accessToken = (String) jsonObject.get("access_token");
```
These should be constants.

tyfranck replied:
fixed

gregorydlogan commented on modules/transcription-service-google-speech-impl/src/main/java/org/opencastproject/transcription/googlespeech/GoogleSpeechTranscriptionService.java:
```diff
@@ -0,0 +1,1064 @@
+  protected void deleteStorageFile(String mpId, String token) throws IOException {
+    CloseableHttpClient httpClientDel = makeHttpClient();
+    GoogleSpeechTranscriptionServiceStorage storage = new GoogleSpeechTranscriptionServiceStorage();
+    storage.deleteGoogleStorageFile(httpClientDel, storageBucket, mpId + ".flac", token);
```
What if the user changes the encoding settings?  Is this always a `.flac`?

tyfranck replied:
Good point I'll put this in the config file and set default as 'flac'

tyfranck replied:
fixed

gregorydlogan commented on modules/transcription-service-google-speech-impl/src/main/java/org/opencastproject/transcription/googlespeech/GoogleSpeechTranscriptionService.java:
```diff
@@ -0,0 +1,1064 @@
+      logger.warn(String.format("Error received for media package %s, job id %s",
+              jobControl.getMediaPackageId(), jobId));
+      // Send notification email
+      sendEmail("Transcription ERROR",
```
I would be tempted to turn most of the parameters for your usage of `sendEmail` into constants.

gregorydlogan commented on modules/transcription-service-google-speech-impl/src/main/java/org/opencastproject/transcription/googlespeech/GoogleSpeechTranscriptionService.java:
```diff
@@ -0,0 +1,1064 @@
+
+      try {
+        // Find jobs that are in progress and jobs that had transcription complete
+        List<TranscriptionJobControl> jobs = database.findByStatus(TranscriptionJobControl.Status.Progress.name(),
```
IMO, `findBy*` which takes `Status` enum types should directly take the type rather than Strings.

gregorydlogan commented on modules/transcription-service-google-speech-impl/src/main/java/org/opencastproject/transcription/googlespeech/GoogleSpeechTranscriptionService.java:
```diff
@@ -0,0 +1,1064 @@
+          String jobId = j.getTranscriptionJobId();
+
+          // If the job in progress, check if it should already have finished.
+          if (TranscriptionJobControl.Status.Progress.name().equals(j.getStatus())) {
```
I would rename `Progress` to be `InProgress`, otherwise this is quite confusing unless read aloud.

tyfranck replied:
fixed

gregorydlogan commented on modules/transcription-service-google-speech-impl/src/main/java/org/opencastproject/transcription/googlespeech/GoogleSpeechTranscriptionService.java:
```diff
@@ -0,0 +1,1064 @@
+          if (TranscriptionJobControl.Status.Progress.name().equals(j.getStatus())) {
+            // If job should already have been completed, try to get the results. Consider a buffer factor so that we
+            // don't try it too early.
+            if (j.getDateCreated().getTime() + j.getTrackDuration() + completionCheckBuffer * 1000 < System
```
Aside: They're saying they take at least real time?  I would have expected that they could process a track faster than real-time....

tyfranck replied:
yes indeed.  I've set it to half of of the track duration

tyfranck replied:
fixed

gregorydlogan commented on modules/transcription-service-google-speech-impl/src/main/java/org/opencastproject/transcription/googlespeech/GoogleSpeechTranscriptionService.java:
```diff
@@ -0,0 +1,1064 @@
+                    database.updateJobControl(jobId, TranscriptionJobControl.Status.Canceled.name());
+                    // Delete file stored on Google storage
+                    String token = getRefreshAccessToken();
+                    deleteStorageFile(mpId, token);
```
We don't want to optionally initiate a retry here?

tyfranck replied:
Possible but not part of the initial release, I'll leave it for the next version. 
It needs couple of tweaks and designed properly..

gregorydlogan commented on modules/transcription-service-google-speech-impl/src/main/java/org/opencastproject/transcription/googlespeech/GoogleSpeechTranscriptionService.java:
```diff
@@ -0,0 +1,1064 @@
+
+            // Apply workflow to attach transcripts
+            Map<String, String> params = new HashMap<String, String>();
+            params.put("transcriptionJobId", jobId);
```
Constant

tyfranck replied:
fixed

gregorydlogan commented on modules/transcription-service-google-speech-impl/src/main/java/org/opencastproject/transcription/googlespeech/GoogleSpeechTranscriptionService.java:
```diff
@@ -0,0 +1,1064 @@
+
+    try {
+      WorkflowDefinition wfDef = workflowService.getWorkflowDefinitionById(wfDefId);
+      Workflows workflows = wfUtil != null ? wfUtil : new Workflows(assetManager, workspace, workflowService);
```
We try and avoid ternaries, although this is not enforced by checkstyles.

tyfranck replied:
fixed

gregorydlogan commented on modules/transcription-service-google-speech-impl/src/main/java/org/opencastproject/transcription/googlespeech/endpoint/GoogleSpeechTranscriptionRestService.java:
```diff
@@ -0,0 +1,130 @@
+    try {
+      jobResult = service.getTranscriptionResults(jobId);
+    } catch (TranscriptionServiceException e) {
+      throw new TranscriptionServiceException("Could not get transcription result", e);
```
IIRC we need to build a Response object still, or we get an ugly stack trace in the browser in this case.  Check some of the other common services to see.  This comment applies to all of the REST endpoints in this PR.

tyfranck replied:
fixed

gregorydlogan commented on modules/transcription-service-persistence/src/main/java/org/opencastproject/transcription/persistence/TranscriptionDatabaseImpl.java:
```diff
@@ -0,0 +1,166 @@
+    if (dtoProvider != null) {
+      return dtoProvider.toTranscriptionProviderControl();
+    }
+    return null;
```
What's the advantage here vs throwing an exception?  If someone's asking for an invalid/missing provider then there's a fairly major error.

tyfranck replied:
The advantage here is that if provider is missing then create it. 
Providers are not created by default so we don't want this throwing an exception/ending the application, we want to create the provider instead and continue the application.

gregorydlogan commented on modules/transcription-service-persistence/src/main/java/org/opencastproject/transcription/persistence/TranscriptionDatabaseImpl.java:
```diff
@@ -0,0 +1,166 @@
+        dtoProvider = TranscriptionProviderControlDto.findIdByProvider(emf.createEntityManager(), provider);
+        return dtoProvider.toTranscriptionProviderControl();
+      }
+      return null; // not found
```
Should be an exception IMO, see comment below.

tyfranck replied:
same reply as above

gregorydlogan commented on modules/transcription-service-persistence/src/main/java/org/opencastproject/transcription/persistence/TranscriptionJobControlDto.java:
```diff
@@ -0,0 +1,246 @@
+  @XmlAttribute
+  private long id;
+
+  @Column(name = "media_package_id", nullable = false, length = 128)
```
`media_package_id` should be `mediapackage_id`.  You may have gotten this from my AWS S3 dist code, or from the Watson code.  Both sets are wrong :D  I'll file a PR shortly.

gregorydlogan commented on modules/transcription-service-persistence/src/main/java/org/opencastproject/transcription/persistence/TranscriptionJobControlDto.java:
```diff
@@ -0,0 +1,246 @@
+  public static List<TranscriptionJobControlDto> findByStatus(EntityManager em, final String... status)
+          throws TranscriptionDatabaseException {
+    Collection<String> statusCol = new HashSet<String>();
+    for (String st : status) {
```
I see what you're doing, and understand why.  I would still break this into two methods, one which takes a single Status, the other which takes a Set<Status>.  The single input would wrap its input into a Set, then call the Set input to keep the logic in one place.

Hint: What if status is `null`?

tyfranck replied:
I don't understand the 2nd sentence. 
The current implementation will handle both single status and a set of statuses.  I've added a check for null status. 
What's the reasoning behind splitting into 2 methods?

gregorydlogan replied:
Right now, if I were to call this method with a null as the final parameter, I would see a null pointer exception on 176 when Java attempts to iterate on a null.  I don't see a null check, but that might be a local change that's not here yet.  If it's checked we can leave it as is.

tyfranck replied:
changed was local, I've now pushed it.

gregorydlogan commented on modules/transcription-service-persistence/src/main/java/org/opencastproject/transcription/persistence/TranscriptionProviderControlDto.java:
```diff
@@ -0,0 +1,134 @@
+      query.setParameter("id", id);
+      return (TranscriptionProviderControlDto) query.getSingleResult();
+    } catch (NoResultException e) {
+      return null; // Not found
```
I would throw a `NotFoundException` here, which we use in a few places for things like this.

tyfranck replied:
Same reply as previously explained

gregorydlogan commented on modules/transcription-service-workflowoperation/src/main/java/org/opencastproject/transcription/workflowoperation/GoogleSpeechAttachTranscriptionOperationHandler.java:
```diff
@@ -0,0 +1,178 @@
+      // Add tags
+      if (targetTagOption != null) {
+        for (String tag : asList(targetTagOption)) {
+          if (StringUtils.trimToNull(tag) != null) {
```
`!StringUtils.isBlank` would be a better fit here.

lkiesow replied:
[`StringUtils.isNotBlank(…)`](https://commons.apache.org/proper/commons-lang/apidocs/org/apache/commons/lang3/StringUtils.html#isNotBlank-java.lang.CharSequence-) ;-)

gregorydlogan replied:
@lkiesow has a better solution :)

tyfranck replied:
fixed

gregorydlogan commented on modules/transcription-service-workflowoperation/src/main/java/org/opencastproject/transcription/workflowoperation/GoogleSpeechStartTranscriptionOperationHandler.java:
```diff
@@ -0,0 +1,185 @@
+      }
+    }
+    if (StringUtils.isNotBlank(langCode)) {
+      language = StringUtils.trim(langCode);
```
`langCode` has already been trimmed to null above.  IMO line 120 should read `language = ...` since `language` as implemented ends up as either the non-blank `langCode` value, or `null`.

tyfranck replied:
fixed

gregorydlogan reviewed (changes requested):
This looks really good, thanks for the hard work @tyfranck.  I know this looks like a ton of complaints, but the majority is style and nitpicks.  Ideally the database bits would be broken out into their own PR, but at this point I've gone through it all so we might as well leave it as is :)

tyfranck commented:
@gregorydlogan thanks for the reviews and comments, really useful. 
I am not clear or don't understand one or two but will get back to you in due course.

smarquard commented:
@tyfranck, I was wondering what the value is of the provider table. Is there a possibility that we may store may provider-related info in this table in future? Otherwise it would seem simpler to just have a string providerid in the transcription jobs table.

tyfranck commented:
@smarquard, yes, the provider table provides the ability to add more infos about providers in the future if needed. It also allows those who sign up to many providers (google, ibm) to easily identify them.  I can see Amazon, Microsoft being added to the list in the future. So having a human readble way to identify them is necessary.
providerId only, I agree is simpler but might be an issue later on when it comes to identify who is who.
I am opened for suggestions though.

gregorydlogan commented:
I really like the idea of the provider table.  We're going to have multiple implementations as soon as this goes in anyway, we should get ahead of the technical debt and do it right the first time :)

KatrinIhler commented:
@tyfranck Conflicts in the database migration script!

lkiesow commented:
Note that this pull request has conflicts in `docs/upgrade/6_to_7/mysql5.sql￼` which should be `docs/upgrade/7_to_8/mysql5.sql￼`anyway since this will be an Opencast 8 feature

tyfranck committed:
Merge branch 'develop' into f/MH-13265_Automate_caption_using_google_speech_API

KatrinIhler commented:
@tyfranck Test failures!

gregorydlogan commented:
@tyfranck, any idea what the ETA for movement on this would be?

tyfranck commented:
@gregorydlogan, working my way through the comments and addressing them when possible. Most of the trivial changes have been done locally. Some comments require more thinking and have knock on effect on other part so need a bit more time to look into them. I'll spend some time on it this week and merge some of the new changes. The plan is to get it done by the end of this month.

tyfranck committed:
Merge branch 'develop' into f/MH-13265_Automate_caption_using_google_speech_API

tyfranck committed:
MH-13265 updated various documentation, classes and methods following review

tyfranck committed:
MH-13265 uncommented enabled=false

tyfranck commented:
@gregorydlogan, apologies for the delay but I've now addressed your comments and pushed all the changes after a bit of fire fighting with various conflict :-).
In addition the pull request have successfully passed all automated checks/build stage.

gregorydlogan commented on modules/caption-impl/src/main/java/org/opencastproject/caption/converters/GoogleSpeechCaptionConverter.java:
```diff
@@ -0,0 +1,189 @@
+    try {
+      // No language to specify so define size of a transcripts line
+      transcriptionLineSize = Integer.parseInt(languageLineSize.trim());
+      logger.info("Transcripts line size {} used", LINE_SIZE);
```
This is logging the default, not the parsed line size

tyfranck replied:
fixed

gregorydlogan commented:
@tyfranck two final nits to pick then we're done!

tyfranck committed:
MH-13265 changed log format and logged correct value
